src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -229,10 +232,11 @@ def finalizer():
           stderr.flush()
         finally:
           time.sleep(.001)  # HACK: Sleep 1ms in the main thread to free the GIL.
-          writer.stop()
+          stdout_pipe.stop_writing()
```

Comment:
So, files/pipes already support a "I'm closed" flag. Would closing the write side (ie, sending EOF) work here?

Reply:
So, if we didn't use the `Pipe` abstraction, that's what we would have to do. However, pipes are not closed when the write end is closed. Instead, the `select` in `NailgunStreamWriter` will just not return the read fd ever. This makes the cases "I haven't read anything yet" and "I have finished reading and the pipe is closed" indistinguishable for `NSW` as it stands before this PR. 

Therefore, the `PipedNailgunStreamWriter` would have to keep track of the write file descriptors to be able to tell when a pipe is closed. That's feasible, but given that we have the `Pipe` abstraction, I thought it best to use that for readability.

Reply:
TIL!